presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/README.md

```diff
@@ -0,0 +1,162 @@
+# Presto planner and optimizer tests
+
+## Quickstart
+
+### The basics
+
+If you want to see examples of basic use of the plan testing framework, look at
+TestPlanFramework.java. If you'd like to try to begin writing new tests, look
+at PlanMatchPattern.java to see what sorts of patterns are already
+well-supported by the framework.
+
+If you're dealing with Expressions, you'll probably want to take a look at
+ExpressionVerifier as well.
+
+### Testing planner code
+
+Planner code should be tested by comparing the unoptimized plan to a set of
+expectations. For various reasons related to how symbols are handled in
+unoptimized plans, this is frequently impossible. The minimum set of optimizers
+that needs to be applied is likely to be some subset of
+
+- UnaliasSymbolReferences
+- PruneIdentityProjections
+- PruneUnreferencedOutputs
+
+`TestQuantifiedComparison` illustrates unit testing planner code.
+
+### Testing optimizer code
+
+Optimizer code will generally need to be subjected to two types of tests:
+
+1. Unit tests on the minimum set of optimizers needed for the optimizer to
+   function.
+2. A test using the full set of optimizers that ensures that the introduction
+   of new optimizers or reordering of existing optimizers prevents the
+   optimizer under test from working correctly.
+
+TestMergeWindows illustrates both types of tests and goes further into the
+rationale for both types of tests.
+
+### Supporting new PlanNodes
+
+There are several things you might want to verify in a PlanNode:
+
+1. The contents of the PlanNode itself (e.g. equi criteria in a JoinNode)
+2. The contents of some Symbol-producing member of the PlanNode (e.g.
+   FunctionCalls in a WindowNode)
+
+These should be handled separately.
+
+#### Fields that do not produce Symbols
+
+Fields that don't produce symbols should be verified in a new Matcher that
+is specific to the PlanNode. As much as possible, the matcher should call
+an object's own .equals method rather than implementing the comparison itself.
+
+See JoinMatcher for an example of comparing non-symbol producing fields.
+
+#### Fields that produce Symbols
+
+Fields that produce Symbols should be compared with an RvalueMatcher that can
+be used with the special `Matcher` `Alias`. `Alias` is used to bind those
+Symbols to an alias so they can be referenced by alias further up the tree. 
+
+Examples of matching fields that produce symbols:
+
+Node type | Matching infrastructure
+----------|------------------------
+TableScanNode | ColumnReference
+ProjectNode | ExpressionAssignment
+AggregationNode | AggregationFunctionMatcher and FunctionCallProvider 
+
+## Aliases and ExpectedValueProviders
+
+At a high level, Presto plans represent data through Symbols. A TableScanNode,
+for instance, generates one symbol per column in the table as its output.
+ProjectNodes can apply a calculation to some input Symbol and generate a new
+Symbol as output.
+
```

Comment:
I'd mention symbol tracking through aliases here, not earlier. So maybe you can move https://github.com/Teradata/presto/pull/417/commits/c43c467cb9e875870cdade078253a0c0cfadb5e4#diff-08700b6c8524843d8b16847b73536343R62 somehow here?